app/gui2/public/visualizations/builtins.ts

```diff
@@ -0,0 +1,53 @@
+import { ref, type Ref } from 'vue'
+
+export interface Vec2 {
+  readonly x: number
+  readonly y: number
+}
+
+export interface RGBA {
+  red: number
+  green: number
+  blue: number
+  alpha: number
+}
+
+export interface Theme {
+  getColorForType(type: string): RGBA
+}
+
+export const DEFAULT_THEME: Theme = {
+  getColorForType(type) {
+    let hash = 0
+    for (const c of type) {
+      hash = 0 | (hash * 31 + c.charCodeAt(0))
+    }
+    if (hash < 0) {
+      hash += 0x80000000
+    }
+    const red = (hash >> 24) / 0x180
+    const green = ((hash >> 16) & 0xff) / 0x180
+    const blue = ((hash >> 8) & 0xff) / 0x180
+    return { red, green, blue, alpha: 1 }
+  },
+}
+
+interface UpdateableRef<T> extends Ref<T> {
+  update(): void
+}
+
+/** A Vue ref which can be updated. */
+export function updateableRef<T>(value: T) {
+  const ref_ = ref({ value })
+  return {
+    get value() {
+      return ref_.value.value
+    },
+    set value(value) {
+      ref_.value = { value }
+    },
+    update() {
+      ref_.value = { value: ref_.value.value }
+    },
+  }
+}
```

Comment:
@Frizi added this to (partially) solve the `watchPostEffect` thing, but it feels quite hacky. thoughts?

Reply:
Oh I understand the issue now. I don't like the solution though. We already have `triggerRef` in `vue` exactly for that purpose. But I would still avoid using it if we can help it.

In this case, you are using those `updateableRef`s only for axes/scales. Looks like d3 axis and scale are keeping all their internal state in local variables, and it's really just a fancy closure. That also tells me we should not keep them in `ref`s in the first place. If anything, those should be computed values, and their inputs should either be props or other refs.

I'm going to show how I imagine that being used in the next comment.